File: src/java/com/churchapp/service/DonationService.java
package com.churchapp.service;

import com.churchapp.entity.Donation;
import com.churchapp.entity.enums.DonationType;
import com.churchapp.repository.DonationRepository;
import io.vavr.control.Option;
import io.vavr.control.Try;
import io.vavr.collection.List;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDate;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional
public class DonationService {
    
    private final DonationRepository donationRepository;
    private final ApplicationEventPublisher eventPublisher;
    
    public Try<Donation> createDonation(Donation donation) {
        return Try.of(() -> {
            log.info("Creating new donation: ${} for member {}", 
                donation.getAmount(), donation.getMember().getId());
            Donation savedDonation = donationRepository.save(donation);
            
            // Publish event for Camel integration
            eventPublisher.publishEvent(new DonationCreatedEvent(savedDonation));
            
            return savedDonation;
        }).onFailure(throwable -> log.error("Failed to create donation", throwable));
    }
    
    public Try<Donation> updateDonation(Donation donation) {
        return Try.of(() -> {
            log.info("Updating donation with id: {}", donation.getId());
            return donationRepository.save(donation);
        }).onFailure(throwable -> log.error("Failed to update donation", throwable));
    }
    
    public List<Donation> getAllDonations() {
        return Try.of(() -> List.ofAll(donationRepository.findAll()))
            .onFailure(throwable -> log.error("Failed to retrieve all donations", throwable))
            .getOrElse(List.empty());
    }
    
    public Option<Donation> getDonationById(Integer id) {
        return Try.of(() -> Option.ofOptional(donationRepository.findById(id)))
            .onFailure(throwable -> log.error("Failed to find donation by id: {}", id, throwable))
            .getOrElse(Option.none());
    }
    
    public List<Donation> getDonationsByMember(Integer memberId) {
        return Try.of(() -> List.ofAll(donationRepository.findByMemberId(memberId)))
            .onFailure(throwable -> log.error("Failed to find donations by member id: {}", memberId, throwable))
            .getOrElse(List.empty());
    }
    
    public List<Donation> getDonationsByType(DonationType donationType) {
        return Try.of(() -> List.ofAll(donationRepository.findByDonationType(donationType)))
            .onFailure(throwable -> log.error("Failed to find donations by type: {}", donationType, throwable))
            .getOrElse(List.empty());
    }
    
    public List<Donation> getDonationsByDateRange(LocalDate startDate, LocalDate endDate) {
        return Try.of(() -> List.ofAll(donationRepository.findByDonationDateBetween(startDate, endDate)))
            .onFailure(throwable -> log.error("Failed to find donations by date range", throwable))
            .getOrElse(List.empty());
    }
    
    public Option<BigDecimal> getTotalDonationsByMember(Integer memberId) {
        return Try.of(() -> Option.of(donationRepository.sumDonationsByMemberId(memberId)))
            .onFailure(throwable -> log.error("Failed to calculate total donations by member: {}", memberId, throwable))
            .getOrElse(Option.none());
    }
    
    public Option<BigDecimal> getTotalDonationsByType(DonationType donationType) {
        return Try.of(() -> Option.of(donationRepository.sumDonationsByType(donationType)))
            .onFailure(throwable -> log.error("Failed to calculate total donations by type: {}", donationType, throwable))
            .getOrElse(Option.none());
    }
    
    public Option<BigDecimal> getTotalDonationsByDateRange(LocalDate startDate, LocalDate endDate) {
        return Try.of(() -> Option.of(donationRepository.sumDonationsByDateRange(startDate, endDate)))
            .onFailure(throwable -> log.error("Failed to calculate total donations by date range", throwable))
            .getOrElse(Option.none());
    }
    
    public Try<Void> deleteDonation(Integer id) {
        return Try.run(() -> {
            log.info("Deleting donation with id: {}", id);
            donationRepository.deleteById(id);
        }).onFailure(throwable -> log.error("Failed to delete donation with id: {}", id, throwable));
    }
    
    // Event class for Camel integration
    public static class DonationCreatedEvent {
        private final Donation donation;
        
        public DonationCreatedEvent(Donation donation) {
            this.donation = donation;
        }
        
        public Donation getDonation() {
            return donation;
        }
    }
}

File: src/kotlin/com/churchapp/service/DonationService.kt
package com.churchapp.service

import arrow.core.Either
import arrow.core.left
import arrow.core.right
import com.churchapp.entity.Donation
import com.churchapp.repository.DonationRepository
import org.slf4j.LoggerFactory
import org.springframework.context.ApplicationEventPublisher
import org.springframework.stereotype.Service
import java.math.BigDecimal
import java.time.LocalDateTime
import java.util.Optional
import java.util.UUID

@Service
class DonationService(
    private val donationRepository: DonationRepository,
    private val eventPublisher: ApplicationEventPublisher,
) {
    private val logger = LoggerFactory.getLogger(DonationService::class.java)

    // Create donation with functional validation
    @org.springframework.transaction.annotation.Transactional
    fun createDonation(donation: Donation): Either<Exception, Donation> =
        try {
            require(donation.amount.compareTo(BigDecimal.ZERO) > 0) { "Donation amount must be positive" }

            val savedDonation = donationRepository.save(donation)
            try {
                eventPublisher.publishEvent(DonationCreatedEvent(savedDonation))
            } catch (eventEx: Exception) {
                logger.error("Failed to publish DonationCreatedEvent", eventEx)
            }
            savedDonation.right()
        } catch (e: Exception) {
            logger.error("Error creating donation", e)
            e.left()
        }

    // Update donation
    fun updateDonation(donation: Donation): Either<Exception, Donation> =
        try {
            require(donation.id != null) { "Donation ID is required for update" }
            require(donation.amount.compareTo(BigDecimal.ZERO) > 0) { "Donation amount must be positive" }

            donationRepository.save(donation).right()
        } catch (e: Exception) {
            logger.error("Error updating donation", e)
            e.left()
        }

    // Get all donations
    fun getAllDonations(): List<Donation> =
        try {
            donationRepository.findAll()
        } catch (e: Exception) {
            logger.error("Error retrieving all donations", e)
            emptyList()
        }

    // Get donation by ID
    fun getDonationById(id: UUID): Optional<Donation> {
        return donationRepository.findById(id)
    }

    // Get donations by member
    fun getDonationsByMember(memberId: UUID): List<Donation> {
        return donationRepository.findByMemberId(memberId)
    }

    fun getTotalDonationsByMember(memberId: UUID): BigDecimal {
        return donationRepository.findByMemberId(memberId)
            .fold(BigDecimal.ZERO) { acc, donation ->
                acc.add(requireNotNull(donation.amount) { "Donation amount is required" })
            }
    }

    // Get donations by date range
    fun getDonationsByDateRange(
        startDate: LocalDateTime,
        endDate: LocalDateTime,
    ): List<Donation> {
        return donationRepository.findByDonationDateBetween(startDate, endDate)
    }

    // Delete donation
    fun deleteDonation(id: UUID): Either<Exception, Unit> =
        try {
            donationRepository.deleteById(id)
            Unit.right()
        } catch (e: Exception) {
            logger.error("Error deleting donation", e)
            e.left()
        }

    // Event for donation creation
    data class DonationCreatedEvent(val donation: Donation)
}
